### PR TITLE
#117 - (refactor) Multi threading Pre-req - Abstract out Store. And fix the hell fire of all test cases breaking

### PR DESCRIPTION
**Pull request by yashs360**
Abstracted out Store as struct and initialise and use in all modules, rather than accessing and mutation of the globally available store.

This is a pre-req for multi threaded store and a step closer to share nothing model.

Tests are addressed for the change.

### CODE REVIEW

File: core/store.go [hunk 1]
```diff
@@ -15,57 +15,58 @@ type WatchEvent struct {
...
 		LastAccessedAt: getCurrentClock(),
 	}
 	if expDurationMs >= 0 {
-		setExpiry(obj, expDurationMs)
+		store.setExpiry(obj, expDurationMs)
 	}
 	return obj
 }
 
+func (store *Store) ResetStore() {
```
Comment (AshwinKul28) [inline core/store.go:53]: Do we need a locking mechanism for the resetStore function? I think we can take all values inside store to initial stage without acquiring locks.

File: core/store.go [hunk 2]
```diff
@@ -15,57 +15,58 @@ type WatchEvent struct {
...
-		WatchChannel = make(chan WatchEvent, config.KeysLimit)
-	}, WithStoreLock(), WithKeypoolLock())
+func NewStore() *Store {
+	s := &Store{
+		store:   make(map[unsafe.Pointer]*Obj),
+		expires: make(map[*Obj]uint64),
+		keypool: make(map[string]unsafe.Pointer),
+	}
+	WatchChannel = make(chan WatchEvent, config.KeysLimit)
+	return s
```
Comment (AshwinKul28) [inline core/store.go:38]:
```
WatchChannel = make(chan WatchEvent, config.KeysLimit)
	return &Store{
		store:   make(map[unsafe.Pointer]*Obj),
		expires: make(map[*Obj]uint64),
		keypool: make(map[string]unsafe.Pointer),
	}
```

No need of additional variable to hold the store in the function scope.

File: main.go [hunk 3]
```diff
@@ -32,14 +32,16 @@ func main() {
 	// If port not available, raise FATAL error
 	serverFD, err := server.FindPortAndBind()
 	if err != nil {
-		log.Fatal(err)
+		log.Error(err)
```
Comment (AshwinKul28) [inline main.go:35]: Do we need to change the behaviour of exiting from program if port is unavailable?

**Comment by gauravsarma1992** [comment]
Since we are already touching majority of the codebase in this PR, I think we should consider moving the `Store` object to a different package altogether, create an interface with the required methods and the actual store implementing the interface. The other code paths shouldn't be able to access the internal elements of the Store directly and should only use the publicly exposed ones. As we move towards implementing our own datastores, this will allow us to experiment quickly on other data stores, limit the usage of quirky patterns for the data store,  and also protect it against any misuse.
cc @soumya-codes

**Comment by AshwinKul28** [comment]
I second @gauravsarma1992 . We can have a separate package in the internal folder to handle the store object. And `WatchChannel` can be separated from the store. There can be a separate interface that can be implemented by any store and that can have all desired method definitions mentioned. This will be kind of a factory design pattern and it will allow us to implement any kind of store quickly.

File: core/store.go (hunk 1 above)
Comment (yashs360) [reply core/store.go:53]: Would we see any inconsistencies when clients read which the reset is still in progress. Eg. Some aributes are reset but others are not?

File: main.go (hunk 3 above)
Comment (yashs360) [reply main.go:35]: Good call. I'll read up a bit more on the difference in behaviours. his was a change in logging libs and the Error should be the equivalent of Fatal from older lib. I'll confirm.

File: core/store.go (hunk 1 above)
Comment (AshwinKul28) [reply core/store.go:53]: I believe resetStore is only for tests and it doesn't get involved in any client operations. 
Screenshot 2024-08-18 at 4 17 05 PM

**Comment by yashs360** [comment]
Interesting idea moving the store. I am keen keen to try it but i don't want to overload the refactor change. 
Atm the store functions are referred on several parts of code and limiting to a subset of functions could take a while. 

I am happy to create a follow up task for this and unblock the multi threaded storage work. What are your thoughts?

File: core/aof.go [hunk 4]
```diff
@@ -101,10 +101,10 @@ func DumpAllAOF() error {
 
 	log.Println("rewriting AOF file at", config.AOFFile)
 
-	storeMutex.RLock()
-	defer storeMutex.RUnlock()
+	store.storeMutex.RLock()
```
Comment (ashwaniYDV) [inline core/aof.go:104]: we can use withLocks here?

File: core/store.go (hunk 1 above)
Comment (yashs360) [reply core/store.go:53]: Good find. Makes me think then should it even in the store.go. Seems more like a utility func for tests.

File: main.go (hunk 3 above)
Comment (yashs360) [reply main.go:35]: I stand corrected. This was an unnecessary change. I will revert this.

File: core/aof.go (hunk 4 above)
Comment (yashs360) [reply core/aof.go]: Yes can do

File: core/store.go [hunk 5]
```diff
@@ -15,57 +15,57 @@ type WatchEvent struct {
...
-	keypool   map[string]unsafe.Pointer
-	WatchList sync.Map // Maps queries to the file descriptors of clients that are watching them.
-)
+type Store struct {
+	store        map[unsafe.Pointer]*Obj
+	expires      map[*Obj]uint64 // Does not need to be thread-safe as it is only accessed by a single thread.
+	keypool      map[string]unsafe.Pointer
+	storeMutex   sync.RWMutex
+	keypoolMutex sync.RWMutex
+	WatchList    sync.Map // Maps queries to the file descriptors of clients that are watching them.
```
Comment (soumya-codes) [inline core/store.go:24]: We are using a mix of sync.Mutex and sync.Map for concurrency control now. How do we plan to refactor this once we move towards no lock?

**Review by soumya-codes — COMMENTED** [review]
@yashs360 thanks for the PR. I have few fundamental questions/doubts on our approach. 

In particular I wanted to ensure a well defined behaviour and get desired atomicity for operations like `MGET` and `MPUT`.

It would be great if we have a deeper discussion tomorrow regarding this.

cc @JyotinderSingh @gauravsarma1992

**Comment by AshwinKul28** [comment]
@soumya-codes let me also know if possible when you guys are discussing. Want to hear your thoughts on this.

File: core/store.go (hunk 5 above)
Comment (yashs360) [reply core/store.go:24]: All actions discussed are moved to their separate tasks. Not addressing as part of this change

**Review by JyotinderSingh — APPROVED** [review]
Thanks for this refactor @yashs360! The changes look good!

**Comment by AshwinKul28** [comment]
No linter issues in such a big PR. Awesome job @yashs360 🥳🚀

**Pull request merged**